File: phantom/core/scheduler.py
"""Central action loop with weighted random selection and gaussian timing."""

from __future__ import annotations

import logging
import threading

from phantom.config.schema import PhantomConfig
from phantom.core.randomization import Randomizer
from phantom.core.stats import Stats
from phantom.simulators.base import BaseSimulator
from phantom.stealth.anti_detection import AntiDetection

log = logging.getLogger(__name__)


class Scheduler:
    """Runs the main simulation loop in a daemon thread."""

    def __init__(
        self,
        config: PhantomConfig,
        simulators: dict[str, BaseSimulator],
        config_lock: threading.Lock,
        stats: Stats | None = None,
    ) -> None:
        self._config = config
        self._simulators = simulators
        self._config_lock = config_lock
        self._running = threading.Event()
        self._stop = threading.Event()
        self._thread: threading.Thread | None = None
        self._anti_detection = AntiDetection()
        self._stats = stats

    @property
    def is_running(self) -> bool:
        return self._running.is_set()

    def start(self) -> None:
        if self._thread and self._thread.is_alive():
            return
        self._stop.clear()
        self._running.set()
        self._thread = threading.Thread(target=self._loop, daemon=True, name="phantom-scheduler")
        self._thread.start()
        if self._stats:
            self._stats.mark_active(True)
        log.info("Scheduler started")

    def stop(self) -> None:
        self._running.clear()
        self._stop.set()
        if self._stats:
            self._stats.mark_active(False)
        if self._thread:
            self._thread.join(timeout=5.0)
            self._thread = None
        log.info("Scheduler stopped")

    def toggle(self) -> bool:
        """Toggle running state. Returns new state."""
        if self._running.is_set():
            self.stop()
            return False
        self.start()
        return True

    def shutdown(self) -> None:
        self.stop()

    def _loop(self) -> None:
        log.info("Simulation loop started")
        while not self._stop.is_set():
            if not self._running.is_set():
                self._stop.wait(timeout=0.5)
                continue

            with self._config_lock:
                weights = self._config.simulator_weights()
                timing = self._config.timing

            if not weights:
                log.warning("No simulators enabled, waiting...")
                self._stop.wait(timeout=2.0)
                continue

            # Check for idle period
            if Randomizer.should_idle():
                idle_time = Randomizer.idle_duration()
                log.debug("Idle period: %.1fs", idle_time)
                self._stop.wait(timeout=idle_time)
                if self._stop.is_set():
                    break
                continue

            # Select and execute a simulator
            options = list(weights.items())
            chosen = Randomizer.weighted_choice(options)

            # Anti-detection: avoid repetitive patterns
            if self._anti_detection.would_be_repetitive(chosen):
                remaining = [(n, w) for n, w in options if n != chosen]
                if remaining:
                    chosen = Randomizer.weighted_choice(remaining)

            sim = self._simulators.get(chosen)
            if sim is None:
                continue

            try:
                sim_config = self._get_sim_config(chosen)
                sim.execute(sim_config)
                self._anti_detection.record(chosen)
                if self._stats:
                    self._stats.record_action(chosen)
            except Exception:
                log.exception("Simulator %s failed", chosen)

            # Wait before next action
            interval = Randomizer.action_interval(
                mean=timing.interval_mean,
                stddev=timing.interval_stddev,
                minimum=timing.interval_min,
            )
            self._stop.wait(timeout=interval)

        log.info("Simulation loop exited")

    def _get_sim_config(self, name: str):
        """Get the config section for a named simulator."""
        with self._config_lock:
            return getattr(self._config, name)

File: phantom-macos/Phantom/Core/Scheduler.swift
// Scheduler.swift — DispatchQueue-based weighted loop

import Foundation
import os.log

private let logger = Logger(subsystem: "com.phantom.app", category: "Scheduler")

final class Scheduler {
    private let configManager: ConfigManager
    private let simulators: [String: Simulator]
    private let antiDetection = AntiDetection()
    private let queue = DispatchQueue(label: "com.phantom.scheduler", qos: .utility)
    private let lock = NSLock()

    private var _isRunning = false
    private var _pausedSims: Set<String> = []
    private var workItem: DispatchWorkItem?

    // Stats
    private(set) var totalActions: Int = 0
    private(set) var actionsBySimulator: [String: Int] = [:]
    private(set) var lastActionName: String?

    /// Called on each action with the simulator name.
    var onAction: ((String) -> Void)?

    var isRunning: Bool {
        lock.lock()
        defer { lock.unlock() }
        return _isRunning
    }

    var pausedSims: Set<String> {
        lock.lock()
        defer { lock.unlock() }
        return _pausedSims
    }

    init(configManager: ConfigManager, simulators: [String: Simulator]) {
        self.configManager = configManager
        self.simulators = simulators
    }

    func start() {
        lock.lock()
        guard !_isRunning else { lock.unlock(); return }
        _isRunning = true
        lock.unlock()
        logger.info("Scheduler started")
        scheduleNext()
    }

    func stop() {
        lock.lock()
        _isRunning = false
        workItem?.cancel()
        workItem = nil
        lock.unlock()
        logger.info("Scheduler stopped")
    }

    func toggle() -> Bool {
        if isRunning {
            stop()
            return false
        } else {
            start()
            return true
        }
    }

    func toggleSimPause(_ name: String) -> Bool {
        lock.lock()
        defer { lock.unlock() }
        if _pausedSims.contains(name) {
            _pausedSims.remove(name)
            return false
        } else {
            _pausedSims.insert(name)
            return true
        }
    }

    func shutdown() {
        stop()
    }

    // MARK: - Private

    private func scheduleNext() {
        lock.lock()
        guard _isRunning else { lock.unlock(); return }
        lock.unlock()

        let item = DispatchWorkItem { [weak self] in
            self?.tick()
        }
        lock.lock()
        workItem = item
        lock.unlock()

        // Schedule immediately for first tick
        queue.async(execute: item)
    }

    private func tick() {
        lock.lock()
        guard _isRunning else { lock.unlock(); return }
        lock.unlock()

        let config = configManager.config

        // Get enabled simulator weights (excluding paused sims)
        lock.lock()
        let paused = _pausedSims
        lock.unlock()

        let weights = config.simulatorWeights().filter { !paused.contains($0.0) }
        if weights.isEmpty {
            logger.warning("No simulators enabled, waiting...")
            scheduleAfter(seconds: 2.0)
            return
        }

        // Check for idle period (10% chance)
        if Randomizer.shouldIdle() {
            let idleTime = Randomizer.idleDuration()
            logger.debug("Idle period: \(idleTime, format: .fixed(precision: 1))s")
            scheduleAfter(seconds: idleTime)
            return
        }

        // Weighted random choice among enabled simulators
        var chosen = Randomizer.weightedChoice(options: weights)

        // Anti-detection check
        if antiDetection.wouldBeRepetitive(chosen) {
            let remaining = weights.filter { $0.0 != chosen }
            if !remaining.isEmpty {
                chosen = Randomizer.weightedChoice(options: remaining)
            }
        }

        guard let sim = simulators[chosen] else {
            scheduleAfter(seconds: 1.0)
            return
        }

        // Execute chosen simulator
        sim.execute(config: config)
        antiDetection.record(chosen)

        lock.lock()
        totalActions += 1
        actionsBySimulator[chosen, default: 0] += 1
        lastActionName = chosen
        lock.unlock()

        onAction?(chosen)

        // Schedule next tick after Gaussian interval
        let interval = Randomizer.actionInterval(
            mean: config.timing.intervalMean,
            stddev: config.timing.intervalStddev,
            minimum: config.timing.intervalMin
        )
        scheduleAfter(seconds: interval)
    }

    private func scheduleAfter(seconds: Double) {
        lock.lock()
        guard _isRunning else { lock.unlock(); return }
        lock.unlock()

        let item = DispatchWorkItem { [weak self] in
            self?.tick()
        }
        lock.lock()
        workItem = item
        lock.unlock()

        queue.asyncAfter(deadline: .now() + seconds, execute: item)
    }
}
